### ISSUE TITLE
Distinct naming of repeat and its instances

### ISSUE THREAD

**Issue by smoyte** [getodk/xforms-spec#141]
The current spec says this:

"A `<repeat>` cannnot have a label child element. To display a label it should be wrapped inside a `<group>` as shown below:"

```
    <group ref="/data/person">
        <label>Person</label>
        <repeat nodeset="/data/person">
            <input ref="/data/person/name">
                <label>Enter name</label>
            </input>
            <input ref="/data/person/relationship">
                <label>Enter relationship</label>
            </input>
        </repeat>
    </group>
```

Which leads to something like this in the jump screen (just replace 'Group' with 'Person'):

![image](https://user-images.githubusercontent.com/702320/29527700-b35f5192-8667-11e7-89ff-b4e221a7d548.png)

Our client badly wants the ability to display different names for each instance on the jump screen. As far as I can see there is no way to do this presently. If you try something like this: 

```
    <group>
      <label><output value="/data/person/name"/></label>
      <repeat nodeset="/data/person">
        <group>
          <input ref="/data/person/name">
            <label>Name</label>
          </input>
        </group>
      </repeat>
    </group>
```

You get this:

![image](https://user-images.githubusercontent.com/702320/29527813-f7150012-8667-11e7-9708-313459bac8b8.png)

which is nonsense.

What I propose is something like this:

```
    <group>
      <label>People</label>
      <repeat nodeset="/data/person">
        <group>
            <label>Person: <output value="/data/person/name"/></label>
            <input ref="/data/person/name">
            <label>Name</label>
          </input>
        </group>
      </repeat>
    </group>
```

Then the jump screen would read:

People
- Person: Joe
- Person: Bob

I don't particularly care what is shown as the group label in the question view, but it seems like `Person: Joe` would make most sense.

If folks don't define a group inside the repeat, then the old behavior applies.

Thoughts?

**Comment by lognaturel** [comment]
The good news is that this is not a spec change per se, just a recommendation for form creators. I would phrase it something like "If a repeat contains a group with a label, that label may be considered the label for individual repeats by ODK XForms clients. This can be helpful for dynamically naming individual repeats based on the data they contain." Then @hooverlunch's example from above can be shown.

Most clients would keep doing what they're doing now and the behavior would stay the same. Repeats without inner groups would behave the same way they do now. The inner group would be something that Collect and other clients that may show an overview of questions and/or repeats could choose to use in their rendering.

I think the big question we need to answer is whether using inner groups in this way could make the experience worse for existing forms. If so, I think we should look at other options that require the form developer to explicitly request naming of individual repeats.

Most forms probably don't have a group inside of a repeat so those wouldn't change. I imagine that if they do contain a group that group's label is appropriate for describing a single instance of a repeat (e.g. "Person" in the case above). The only case I can think of that would be unfortunate is if a repeat contains several groups. For example:
```xml
<group>
  <label>People</label>
  <repeat nodeset="/data/people">
    <group>
      <label>Demographics</label>
      <input ref="/data/people/demographics/age">
        <label>Age</label>
      </input>
    </group>
    <group>
      <label>Schooling</label>
    </group>
  </repeat>
</group>
```
Using the immediately nested group would mean ending up with `Demographics (1)`, `Demographics (2)`, etc as the names of the repeats. That does not seem ideal.

@yanokwa any idea how common that is? @MartijnR any chance Enketo would provide the option to collapse repeats if they could be named meaningfully?

**Comment by smoyte** [comment]
@lognaturel Good point about the multiple groups inside a repeat. That seems like it could be fairly common.

What if we make it even simpler and say that if a `<repeat>` has `<label>` as a child, we treat that as the instance label? There is a nice parallelism there:

* a `label` as a `group`'s child is the group's label (this is already the case)
* a `label` as a `repeat`'s child is the instance's label (this is what we'd add)

So:

```
    <group>
      <label>People</label>
      <repeat nodeset="/data/person">
        <label>Person: <output value="/data/person/name"/></label>
        <input ref="/data/person/name">
          <label>Name</label>
        </input>
      </repeat>
    </group>
```

**Comment by lognaturel** [comment]
I don't know how bad the multiple groups inside a repeat case really is. In the specific one I shared, is "Demographics 1" really all that worse than "Person 1"? Not having to make a spec change and corresponding changes to the various form generators, form rendering libraries, etc definitely has a lot of advantages.

It seems like not allowing a `label` in a `repeat` was a deliberate choice and I wonder what that was about. Adding it seems ok but it would take a bigger effort to get support for it across the ecosystem. I'm inclined to consider this option only if the nested group option really is totally unacceptable.

**Comment by smoyte** [comment]
Checking back in here... I am fine with either of these approaches (label in repeat or label of first group). What are the next steps here?

**Comment by lognaturel** [comment]
I think that for proposals that may change existing behavior, the standard for approval needs to be even higher than when adding new behavior.

Can you think of other folks who you could bring into this discussion to verify that this would make sense for their forms and/or clients? Are there other proposals that you could comment on or contribute to to encourage others to look at yours? Are there publicly-available forms you could check the assumptions we've made in this thread against? Do you want to facilitate a dev call all about groups and repeats (if so, just comment [here](https://forum.opendatakit.org/t/odk-1-0-developer-calls/7094) with a rough agenda and/or we can hash out details in Slack)?

**Comment by smoyte** [comment]
> Are there publicly-available forms you could check the assumptions we've made in this thread against? 

I like this idea. Should I use the example forms repository? Is that considered like an official 'corpus' of forms against which stuff should be tested? If not, maybe it should, or we should set one up elsewhere?

> Do you want to facilitate a dev call all about groups and repeats (if so, just comment here with a rough agenda and/or we can hash out details in Slack)?

Done.

**Comment by MartijnR** [comment]
I haven't been successful in finding out why a repeat cannot have a label (and also have not seen examples where it has). 

In any case, the label of the first group inside a repeat is fine with me. Although it seems a little verbose, the added advantage is that this could be done without XLSForm syntax additions (because 'begin repeat' adds the label to the group). 

> @MartijnR any chance Enketo would provide the option to collapse repeats

Yes, I think both repeats and groups should be collapsible. This work is being planned now. UPDATE: during implementation I realized only groups need to be collapsible, since repeats created with XLSForm always have a group parent anyway. Made it much easier to implement.

**Comment by lognaturel** [comment]
Thanks @MartijnR, that's very helpful.

Good to discuss this a bit on the latest developer call, thanks for leading @hooverlunch. @grzesiek2010, @ronna could you please think through this one and see if you can identify any other risks? I've also requested some forms with repeats from partners to see whether this would affect them.

**Comment by grzesiek2010** [comment]
@lognaturel @hooverlunch @MartijnR 

At first glance it was strange to me why you want to add additional group. It seemed to my like a workaround and I thought we should do something else (maybe force repeats to to have labels).
But then I realize that if we use repeat group with `field-list` appearance an additional group is added anyway.
I used https://build.opendatakit.org/

Normal repeat group:

```html
  <h:body>
    <group>
      <label ref="jr:itext('/data/untitled1:label')"/>
      <repeat nodeset="/data/untitled1">
      </repeat>
    </group>
  </h:body>
```
Repeat group with `field-list` appearance:
```html
  <h:body>
    <group>
      <label ref="jr:itext('/data/untitled1:label')"/>
      <repeat nodeset="/data/untitled1">
        <group appearance="field-list">
        </group>
      </repeat>
    </group>
  </h:body>
```

So now adding additional group makes sense to me and maybe it's even more consistent (?) since we would have that (additional) group in both cases.

@lognaturel you said

> The only case I can think of that would be unfortunate is if a repeat contains several groups. For example:

It's not a problem you can add as many groups inside as you want but in your example you should add one additional "main" group. It was:

```html
<group>
  <label>People</label>
  <repeat nodeset="/data/people">
    <group>
      <label>Demographics</label>
      <input ref="/data/people/demographics/age">
        <label>Age</label>
      </input>
    </group>
    <group>
      <label>Schooling</label>
    </group>
  </repeat>
</group>
```
but it should be:
```html
<group>
  <label>People</label>
  <repeat nodeset="/data/people">
     <group>
        <label>Something</label>
        <group>
            <label>Demographics</label>
            <input ref="/data/people/demographics/age">
                <label>Age</label>
            </input>
         </group>
         <group>
         <label>Schooling</label>
         </group>
     </group>
  </repeat>
</group>
```
and it's related only to normal repeat groups since we can't add inner groups if your repeat has `field-list` appearance there is an issue: https://github.com/opendatakit/collect/issues/1255

I started testing of course to be sure it's ok and I prepared a small apk if someone want to cover my tests:
[apk.zip](https://github.com/opendatakit/xforms-spec/files/1371126/apk.zip)
(it's only a test apk and the solution is not considered well so it might be something wrong there)

When it comes to the code you can take a look here: https://github.com/grzesiek2010/collect/commit/36f761e579fe1be931f81f7821e8ccd7f875accf

The good news is that old forms work in the same way as before and the changes I implemented affect only new forms.
Here is a list of forms I tested:
[repeatGroupFieldListNew.txt](https://github.com/opendatakit/xforms-spec/files/1371209/repeatGroupFieldListNew.txt)
[repeatGroupFieldListOriginal.txt](https://github.com/opendatakit/xforms-spec/files/1371210/repeatGroupFieldListOriginal.txt)
[repeatGroupMultiGroupsInsideNew.txt](https://github.com/opendatakit/xforms-spec/files/1371211/repeatGroupMultiGroupsInsideNew.txt)
[repeatGroupMultiGroupsInsideOriginal.txt](https://github.com/opendatakit/xforms-spec/files/1371212/repeatGroupMultiGroupsInsideOriginal.txt)
[repeatGroupNew.txt](https://github.com/opendatakit/xforms-spec/files/1371213/repeatGroupNew.txt)
[repeatGroupOriginal.txt](https://github.com/opendatakit/xforms-spec/files/1371214/repeatGroupOriginal.txt)

To sum up: this approach makes sense to me and I can't see any dangerous cases.

**Comment by smoyte** [comment]
Very cool @grzesiek2010!! Thanks for doing all this!

**Comment by lognaturel** [comment]
Thanks for putting together an example, @grzesiek2010, it's always nice to see things in action. 

You're absolutely right that in my "Demographics" example another group could be added. For someone who already had a form in use with the structure described, however, the changed behavior could be a bit surprising. But I think the benefits outweigh the costs. Since @MartijnR is on board and no object has been raised, I think we should proceed.

**Comment by MartijnR** [comment]
Shall we close this by adding a line to the Repeats section that says something like:

"When a client needs to compactly show a single repeat instance in the UI (e.g. as a collapsed repeat or a table-of-contents item), it is recommended to show the label of the first child group of that repeat."

**Comment by lognaturel** [comment]
@MartijnR That sounds perfect to me.
